sklearn_pandas/categorical_imputer.py

```diff
@@ -1,58 +1,77 @@
-"""
-
-Impute missing values from a categorical/string np.ndarray or pd.Series with the most frequent value on the training data.
-
-"""
-
 import pandas as pd
 import numpy as np
 
-from sklearn.base import TransformerMixin
+from collections import Counter
 
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.utils.validation import check_is_fitted
 
-class CategoricalImputer(TransformerMixin):
 
+def _get_mask(X, value):
     """
+    Compute the boolean mask X == missing_values.
+    """
+    if value == "NaN" or \
+       value is None or \
+       (isinstance(value, float) and np.isnan(value)):
+        return pd.isnull(X)
+    else:
+        return X == value
 
-    Attributes
+
+class CategoricalImputer(BaseEstimator, TransformerMixin):
+    """
+    Impute missing values from a categorical/string np.ndarray or pd.Series
+    with the most frequent value on the training data.
+
+    Parameters
     ----------
+    missing_values : string or "NaN", optional (default="NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. None and np.nan are treated
+        as being the same, use the string value "NaN" for them.
+
+    copy : boolean, optional (default=True)
+        If True, a copy of X will be created.
 
-    fill : str
+    Attributes
+    ----------
+    fill_ : str
         Most frequent value of the training data.
 
     """
 
-    def __init__(self):
-
-        self.fill = None
-
-    def fit(self, X):
+    def __init__(self, missing_values='NaN', copy=True):
+        self.missing_values = missing_values
+        self.copy = copy
 
+    def fit(self, X, y=None):
         """
-
         Get the most frequent value.
 
         Parameters
         ----------
             X : np.ndarray or pd.Series
                 Training data.
 
+            y : Passthrough for ``Pipeline`` compatibility.
+
         Returns
         -------
-        CategoricalImputer
-            Itself.
-
+            self: CategoricalImputer
         """
 
-        self.fill = pd.Series(X).mode().values[0]
+        mask = _get_mask(X, self.missing_values)
+        X = X[~mask]
+
+        self.fill_ = Counter(X).most_common(1)[0][0]
```

Comment:
Interesting, for long arrays `Counter` is slower than `series.mode`, but for short arrays is significantly faster:

```
x = pd.Series(['a', 'b', 'c', 'a']* int(1e6))
    ...: 
    ...: %timeit Counter(x).most_common(1)[0][0]
    ...: %timeit x.mode()
    ...: 

1 loop, best of 3: 227 ms per loop
10 loops, best of 3: 141 ms per loop
```

```
x = pd.Series(['a', 'b', 'c', 'a']* int(1e2))
    ...: 
    ...: %timeit Counter(x).most_common(1)[0][0]
    ...: %timeit x.mode()
    ...: 
The slowest run took 4.14 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 37.2 µs per loop
10000 loops, best of 3: 146 µs per loop
```

The break point seems to be around 1e3 elements and I expect many datasets to be as big as that, so let's better use the `mode` method. :)